### PR TITLE
opentracing-php full tree

### PR DESCRIPTION
**Pull request by wladekb**
Place for feedback for initial opentracing implementation in PHP.

PROOF OF CONCEPT

### CODE REVIEW

**Comment by wladekb** [comment]
@bensigelman wrote:
the short version of my high-level feedback in case it helps reduce a round-trip between us:
- thanks. already almost all of the way there IMO, great work.
- https://github.com/Wikia/opentracing-php/blob/master/src/OpenTracing/Span.php#L119 doesn’t need to be part of the Span interface in the other languages… it can be a (static) global function that takes the parent span and the span.tracer() combined with Tracer.startSpan() get the job done.
- the OT interface needs to define a few standard required formats and their respective carriers, much like https://github.com/opentracing/opentracing-go/blob/master/propagation.go#L60 and below
- it’s great that you took a stab at a standard tracer impl, but we are planning to move such helpers out of the main OT repos… so that code may live on, but probably not as a subdir.

oh: and if we move this into the opentracing org it would be good to make it “Copyright 2016: The OpenTracing Authors”. Is the MIT license “standard” for PHP OSS, btw? (Fine by me, but we are trying to choose whatever “reasonable” (MIT, Apache, etc) license is most popular for the particular platform)

**Comment by bhs** [comment]
@wladekb: let me know if you want more detail about any of the above bullets... in general it really pretty reasonable to me and most of the feedback will probably have to come from someone who understands PHP best practices better than I do.

File: src/OpenTracing/Extractor.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Typically the carrier has just arrived across a process boundary, often via
+ * an RPC (though message queues and other IPC mechanisms are also reasonable
+ * places to use an Extractor).
+ *
+ * See Injector and Tracer.extractor().
+ *
+ * @package OpenTracing
+ */
+interface Extractor {
```
Comment (bhs) [inline src/OpenTracing/Extractor.php]: We put Extractor and Injector into a single file in .py and I thought that made sense there – not sure if PHP usually does one public class per file or what.

**Comment by wladekb** [comment]
@gabrys @macbre @mixth-sense @kvas-damian wanna take a look in your spare time?

File: src/OpenTracing/Recorder.php [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace OpenTracing;
+
+interface Recorder {
```
Comment (bhs) [inline src/OpenTracing/Recorder.php]: this probably shouldn't be part of OT proper

File: src/OpenTracing/Standard/Format.php [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace OpenTracing\Standard;
+
+class Format {
```
Comment (bhs) [inline src/OpenTracing/Standard/Format.php]: we will definitely want comments about the appropriate carrier for each format at some point

File: src/OpenTracing/Extractor.php (hunk 1 above)
Comment (wladekb) [inline src/OpenTracing/Extractor.php]: Modern PHP follows the Java-like practise to put each class/interface in a separate file so I think this should stay as it is.

File: src/OpenTracing/Standard/Format.php [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace OpenTracing\Standard;
```
Comment (bhs) [inline src/OpenTracing/Standard/Format.php]: I think that the builtin formats should be part of OT proper, not `Opentracing\Standard`

File: src/OpenTracing/Standard/Format.php [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+class Format {
+
+	const SPLIT_TEXT = 'text';
```
Comment (bhs) [inline src/OpenTracing/Standard/Format.php]: we may have non-split text in the future, so `'split_text'`, etc, likely makes more sense.

File: src/OpenTracing/Standard/PackedHttpHeadersPropagator.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace OpenTracing\Standard;
+
+use OpenTracing;
+
+class PackedHttpHeadersPropagator extends Propagator {
+
+	const HTTP_HEADER_STATE_LOWER = 'x-opentracing-state';
```
Comment (bhs) [inline src/OpenTracing/Standard/PackedHttpHeadersPropagator.php]: the `X` prefix is deprecated for HTTP headers

File: src/OpenTracing/Standard/Tracer.php [hunk 7]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace OpenTracing\Standard;
+
+use OpenTracing;
+
+class Tracer implements OpenTracing\Tracer {
```
Comment (bhs) [inline src/OpenTracing/Standard/Tracer.php]: (per my higher-level comment, the "standard" stuff most likely belongs in its own package as a reference impl... see also the extremely early (i.e., hours old) work at https://github.com/opentracing/basictracer-go. Ideally the "basic" impl for PHP would interoperate with the basic impl for Go (etc) at the wire level.

**Comment by bhs** [comment]
Two other things while I'm thinking of them:
- the various OT libraries have provided access to a global singleton `Tracer` instance... I don't see that here but we need some sort of story on that front
- I don't know much about PHP, but my understanding is that it's very difficult to retain state across requests... not sure if that will affect the way that Tracer/Span are designed, but it might be useful to look at other instrumentation libs in php for example patterns

File: src/OpenTracing/Span.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	public function setOperationName( $operationName );
+
+	/**
+	 * Indicates that the work represented by this span has been completed
+	 * or terminated, and is ready to be sent to the Reporter.
+	 *
+	 * If any tags / logs need to be added to the span, it should be done
+	 * before calling finish(), otherwise they may be ignored.
+	 *
+	 * @param int $finishTime
```
Comment (yurishkuro) [inline src/OpenTracing/Span.php]: Q: does php have a commonly used type for timestamps, like http://php.net/manual/en/class.datetimeinterface.php?  If it must be integer, we need to clarify the format of the value, e.g. microseconds since Unix epoch, etc.

File: src/OpenTracing/Standard/Format.php [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+	const SPLIT_BINARY = 'binary';
+	const PACKED_HTTP_HEADERS = 'packed_http_headers';
+	const RAW_HTTP_HEADERS = 'raw_http_headers';
```
Comment (yurishkuro) [inline src/OpenTracing/Standard/Format.php]: it doesn't look like these two formats should be in the official Formats enum. The reason we had one in Go was because Go has a widely used http.Header type in standard library. But these two formats do not introduce a new intermediate representation, they are simply built on top of split text representation.

The can be provided as util methods. Things placed in the standard Format enum require all tracers to implement them, which is not necessary here.

**Comment by yurishkuro** [comment]
Agree with @bensigelman that `standardtracer` should not be part of this repo. Could rename it to `basictracer` and move to another repo.

It's important to have a no-op as default implementation, which is what `stub` package is doing. To validate API adherence I would recommend adding an API test harness like this one: https://github.com/opentracing/opentracing-python/blob/master/opentracing/harness/api_check.py

**Comment by yurishkuro** [comment]
One more thing - in both Go and Python we ended up defining explicit types for the two standard carriers. I am not seeing them here.

File: src/OpenTracing/Span.php (hunk 8 above)
Comment (wladekb) [inline src/OpenTracing/Span.php]: Basically for performance-related stuff you use `float` received from `microtime(true)` (microsecond precision), and it points out I should've put `float` here instead of `int`.

File: src/OpenTracing/Basic/Propagator.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+	const FIELD_SPAN_ID = 'spanid';
+
+	protected $tracer = null;
+
+	public function __construct( Tracer $tracer ) {
+		$this->tracer = $tracer;
+	}
+
+	protected function validateSpan( OpenTracing\Span $span ) {
+		if ( !$span instanceof Span ) {
```
Comment (macbre) [inline src/OpenTracing/Basic/Propagator.php]: Isn't specifying object class in method declaration enough here?

Comment (wladekb) [inline src/OpenTracing/Basic/Propagator.php]: Unfortunately it would break the OO compatibility between `OpenTracing\Span` interface and the particular implementation eg. this one `OpenTracing\Basic\Span`. Thus we need to do this check at runtime. 
However let me know if you have other idea for solving this compatibility issue.

File: src/OpenTracing/Recorder.php (hunk 2 above)
Comment (wladekb) [inline src/OpenTracing/Recorder.php]: Moved to `Basic`.

File: src/OpenTracing/Standard/Format.php (hunk 4 above)
Comment (wladekb) [inline src/OpenTracing/Standard/Format.php]: Done.

File: src/OpenTracing/Standard/Format.php (hunk 5 above)
Comment (wladekb) [inline src/OpenTracing/Standard/Format.php]: Changed.

File: src/OpenTracing/Standard/PackedHttpHeadersPropagator.php (hunk 6 above)
Comment (wladekb) [inline src/OpenTracing/Standard/PackedHttpHeadersPropagator.php]: Fixed.

File: src/OpenTracing/Span.php (hunk 8 above)
Comment (wladekb) [inline src/OpenTracing/Span.php]: Changed to `float` in all places. The unit is standard unit in PHP: seconds.

**Comment by wladekb** [comment]
@yurishkuro 
1. I plan to add tests after the API gets a bit more stable.
2. I'll move the Basic tracer to separate repo later.
3. What's the reason for explicit types for carriers?
@bensigelman 
yes, in PHP every request starts from scratch unless you create a C extension that has different life cycle
I could add a `TracerSingleton` class with `getInstance` (automatically creates a `Stub` instance if none was set before) and `setInstance` methods that would emulate a singleton, but on the other hand you would still need to load the chosen `Tracer` implementation during bootstrap of your app.

**Comment by beberlei** [comment]
Global Tracer could mimic this: https://github.com/opentracing/opentracing-go/blob/master/globaltracer.go

Having a class `OpenTracing\GlobalTracer` with `public static function init(Tracer $tracer)`, `public static function get()` and `public static startSpan($operationName);`.

**Comment by beberlei** [comment]
About `Span::startChild`, I see only three ways here:
- Make Span an abstact class and add the `startChild` implementation as static final function.
- Move startChild into a function in OpenTracing namespace and use Composer's `file` autoload strategy to include the file with the function during every request (not recommended).
- Introduce a static class with utility functions, maybe even the code for GlobalTracer as well, `OpenTracing\Utils` for example. But this would introduce a naming problem.

**Comment by wladekb** [comment]
@beberlei I started with creating an abstract class initially but then decided to go with interfaces since you don't have multiple inheritance in PHP and I didn't want to limit implementors. However given those 3 options i would call the abstract class approach as the most clean. How about defining a `trait` that would implement just this single method?

**Comment by beberlei** [comment]
@wladekb I like abstract class the most as well, a trait cannot be called statically and it has no interface, so users cannot trust objects to implement methods of a trait. This makes it bad to use in a generic library imho. If we have:

``` php
abstract class Span
{
    public static final function startChild(Span $parent, $operationName)
    {
        return $parent->getTracer()->startSpan($operationName, $parent);
    }
}
```

Or we do it as a dynamic method:

``` php
abstract class Span
{
    public final function startChild($operationName)
    {
        return $this->getTracer()->startSpan($operationName, $this);
    }
}
```

I prefer the second, it deviates a little from Go and Python, but is more idomatic in PHP.

In this case we could apply the same to `Tracer` interface and introduce the global tracer methods on an abstract `Tracer` class instead of introducing a second.

I don't think multiple inheritance is needed in implementations here, they are usually just adapters anyways, i dont see the value for implementors having interfaces over abstract classes here.

**Comment by wladekb** [comment]
@beberlei I think modern PHP is OO-based and thus to create a new `Span` object you need to derive it from a `Tracer` or other `Span` object so both of them should have (probably identical) non-static methods to achieve that. I can't fit a static method/function to this mindset so I'd go with creating a dynamic method. Makes sense?

**Comment by beberlei** [comment]
@wladekb My second example had a copy paste error, the static was wrong there. So i agree yes, we should have `$span->startChild($operationName);`, second example.

File: src/OpenTracing/Basic/Span.php [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /**
+     * Indicates that the work represented by this span has been completed
+     * or terminated, and is ready to be sent to the Reporter.
+     *
+     * If any tags / logs need to be added to the span, it should be done
+     * before calling finish(), otherwise they may be ignored.
+     *
+     * @param float $finishTime
```
Comment (yurishkuro) [inline src/OpenTracing/Basic/Span.php]: please clarify the precision of this value in the comments (e.g. seconds since Unit epoch or whichever the convention is), and reference the sys function used to obtain it.

File: src/OpenTracing/Tracer.php [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * Gets current GlobalTracer singleton.
+     *
+     * Returns null if not initialized yet.
+     *
+     * @return Tracer|null
+     */
+    public static final function getGlobalTracer()
+    {
+        return self::$globalTracerInstance;
```
Comment (beberlei) [inline src/OpenTracing/Tracer.php:96]: Should this create a noop tracer when empty? Not sure about the other libraries behaviors and if this is standardized.

File: src/OpenTracing/Span.php [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+     * @return mixed
+     */
+    abstract public function getTraceAttribute($key);
+
+    /**
+     * Provides access to the Tracer that created this Span.
+     *
+     * @return Tracer
+     */
+    public function getTracer() {
```
Comment (beberlei) [inline src/OpenTracing/Span.php]: I think this should be abstract as well, otherwise its forcing `$this->tracer` to exist, maybe implementations don't need that.

File: src/OpenTracing/Tag.php [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+namespace OpenTracing;
+
+class Tag
+{
+
+    const PEER_SERVICE = "peer.service";
+    const PEER_HOSTNAME = "peer.hostname";
+    const PEER_HOST_IPV4 = "peer.ipv4";
+    const PEER_HOST_IPV6 = "peer.ipv6";
+    const PEER_PORT = "peer.port";
```
Comment (yurishkuro) [inline src/OpenTracing/Tag.php:12]: I don't recall how good PHP is with classes and type validation, if it can catch passing the wrong value type into a method (e.g. string to method expecting integer), then instead of string tags it's better to use the pattern from Go.

**Comment by yurishkuro** [comment]
> 1. What's the reason for explicit types for carriers?

so that both the end user and tracer implementor know what they are dealing with. At the moment split text propagator treats the carrier as a flat dictionary - that means it's not really _split_, the end user has no way of deciding which keys from that dictionary represent trace context vs. baggage.

File: src/OpenTracing/Tag.php (hunk 14 above)
Comment (wladekb) [inline src/OpenTracing/Tag.php:12]: PHP 5 can assure the argument is an instance of a given class (or its subclass)
PHP 7 can assure the scalar types for arguments too (it's released for a few months now so we cannot aim at this version yet)

File: src/OpenTracing/Tracer.php (hunk 12 above)
Comment (wladekb) [inline src/OpenTracing/Tracer.php:96]: That's what I coded at the beginning but then reverted as I didn't find any reference for it.

**Comment by wladekb** [comment]
@yurishkuro would an `array` be enough for specifying the carrier type? (`array` is like a `map` in Go but as PHP is weak-typed lang you don't need to specify types of keys and values)

**Comment by yurishkuro** [comment]
If you look at Go/Python impls, the carrier type is defined conceptually as a struct with two maps (or two binary blobs). Is there a reason why using an (associative) array in PHP is better than an explicit type?

File: src/OpenTracing/Span.php (hunk 13 above)
Comment (wladekb) [inline src/OpenTracing/Span.php]: Done.

**Comment by wladekb** [comment]
@yurishkuro In PHP you usually go with plain arrays or full-blown classes that do something. Anyway for the sake of consistency I updated to have specific carrier types for SplitText and SplitBinary.

**Comment by bhs** [comment]
@wladekb:

> @bensigelman 
> 
> yes, in PHP every request starts from scratch unless you create a C extension that has different life cycle
> 
> I could add a TracerSingleton class with getInstance (automatically creates a Stub instance if none was set before) and setInstance methods that would emulate a singleton, but on the other hand you would still need to load the chosen Tracer implementation during bootstrap of your app.

@bcronin would you mind taking a quick look at this and comparing against your PHP instrumentation trauma[s]?

**Comment by wladekb** [comment]
@bensigelman @bcronin FYI, we're a few hours of work later and there's a `Tracer::getGlobalTracer` and `Tracer::setGlobalTracer` but it currently doesn't autospawn a Stub implementation

File: README.md [hunk 15]
```diff
@@ -0,0 +1,19 @@
+OpenTracing API for PHP
+=======================
+
+This package is an unofficial PHP platform API for OpenTracing.
+
+## Required Reading
+
+In order to understand the Go platform API, one must first be familiar with the
```
Comment (bhs) [inline README.md:8]: s/Go/PHP/

**Comment by bhs** [comment]
@wladekb can you summarize the status here? Are you just waiting for an LGTM or are there significant open issues? Thanks.

File: src/OpenTracing/Tracer.php [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * Returns an Injector instance corresponding to $format
+     *
+     * @throws InvalidFormatException
+     *
+     * @param string $format
+     * @return Injector
+     */
+    abstract public function injector($format);
```
Comment (bhs) [inline src/OpenTracing/Tracer.php]: FYI, the spec has been simplified to remove the Injector/Extractor concept and move `Inject`/`Join` into `Tracer` itself... https://github.com/opentracing/opentracing.github.io/issues/70

Here are the diffs in [Go](https://github.com/opentracing/opentracing-go/pull/65) and [Python](https://github.com/opentracing/opentracing-python/pull/20).

**Comment by bhs** [comment]
@wladekb I've been talking with a PHP-using org that is interesting in OT and I wasn't sure of your timeline for picking this up again. Anything to share? And/or would you be irritated/annoyed if it got forked and continued elsewhere from a dev standpoint? You have done some great work here and it looks like it's getting close to where it needs to be.

**Comment by wladekb** [comment]
@bensigelman 

I'd say this is a fake PR as it tried to merge _from_ _master_ to an arbitrary empty branch. I created it to make browsing the code easier and have all files in a single place - the "Files changed" tab and make sure we have a common place to place general comments.

As I had to move my full-time focus away from opentracing I thought to resync the API in batches. I've been keeping track briefly and have seen the evolution is ongoing so I still waited for a best moment. If you think that now is the moment just let me know and I'll update it within a couple of days.

That's a very good news that some other companies are getting interested in opentracing. I fully believe in the goals and I'm happy to have seen the progress in the project over the last couple of weeks. However I lost an ability to focus on it full-time as we decided not to use the opentracing API yet I think it makes sense to fork/move it to opentracing org. Let me know if you need any cooperation from me.

**Comment by bhs** [comment]
@wladekb thanks for the update! We are getting close to stabilization with respect to the propagation specification. The only other important issues in my mind are info/error logging details (which you've brought up yourself IIRC) and how to set something like an error bit on a Span instance, but both of those will likely be strict additions to the API, not changes.
